File: packages/app-floorplan/src/handlers.test.ts
import { describe, expect, it } from "vitest";
import {
  FloorPlanDocument,
  LengthMm,
  PointMm,
} from "@toporoom/domain-floorplan";
import { FakeGeometryPort, InMemoryDocumentStore } from "@toporoom/ports/fakes";
import { AddOpeningHandler } from "./add-opening-handler.js";
import { AddWallHandler } from "./add-wall-handler.js";
import { GeometryRebuildPolicy } from "./rebuild-policy.js";
import { SetMeasurementHandler } from "./set-measurement-handler.js";

async function setup() {
  const store = new InMemoryDocumentStore();
  const geometry = new FakeGeometryPort();
  const rebuild = new GeometryRebuildPolicy(geometry);
  const doc = FloorPlanDocument.create({ id: "doc_1" });
  await store.save(doc.toSceneIR());
  return {
    store,
    geometry,
    addWall: new AddWallHandler(store, rebuild),
    addOpening: new AddOpeningHandler(store, rebuild),
    setMeasurement: new SetMeasurementHandler(store, rebuild),
    storeyId: doc.storeys[0]!.id,
  };
}

describe("floorplan command handlers", () => {
  it("AddWall mutates the document, emits WallAdded, and rebuilds", async () => {
    const ctx = await setup();
    const result = await ctx.addWall.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      start: { x: 0, y: 0 },
      end: { x: 4000, y: 0 },
      thicknessMm: 200,
      heightMm: 2800,
      kind: "exterior",
    });
    expect(result.events.map((e) => e.type)).toContain("WallAdded");
    expect(result.rebuild.ok).toBe(true);
    expect(ctx.geometry.lastBuildRequest?.semantics.documentId).toBe("doc_1");
    const saved = await ctx.store.load("doc_1");
    expect(saved?.storeys[0]?.walls).toHaveLength(1);
  });

  it("AddOpening mutates the host wall and emits OpeningAdded", async () => {
    const ctx = await setup();
    await ctx.addWall.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      start: { x: 0, y: 0 },
      end: { x: 4000, y: 0 },
      thicknessMm: 200,
      heightMm: 2800,
      kind: "exterior",
    });
    const result = await ctx.addOpening.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      openingId: "op_door",
      kind: "door",
      widthMm: 900,
      heightMm: 2100,
      offsetMm: 800,
      sillHeightMm: 0,
    });
    expect(result.events.map((e) => e.type)).toContain("OpeningAdded");
    const saved = await ctx.store.load("doc_1");
    expect(saved?.storeys[0]?.walls[0]?.openings[0]?.id).toBe("op_door");
  });

  it("SetMeasurement writes laser source onto SceneIR and resizes the opening", async () => {
    const ctx = await setup();
    await ctx.addWall.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      start: { x: 0, y: 0 },
      end: { x: 4000, y: 0 },
      thicknessMm: 200,
      heightMm: 2800,
      kind: "exterior",
    });
    await ctx.addOpening.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      openingId: "op_door",
      kind: "door",
      widthMm: 800,
      heightMm: 2100,
      offsetMm: 800,
      sillHeightMm: 0,
    });
    const result = await ctx.setMeasurement.execute({
      documentId: "doc_1",
      measurementId: "m_door",
      kind: "length",
      valueMm: 900,
      source: "laser",
      instrumentId: "laser_sku_x",
      target: { entityType: "opening", entityId: "op_door", field: "width" },
    });
    expect(result.events.map((e) => e.type)).toContain(
      "FloorPlanSemanticsChanged",
    );
    const saved = await ctx.store.load("doc_1");
    expect(saved?.measurements[0]).toMatchObject({
      source: "laser",
      valueMm: 900,
    });
    expect(saved?.storeys[0]?.walls[0]?.openings[0]?.widthMm).toBe(900);
  });

  it("RebuildPolicy calls GeometryPort when semantics change", async () => {
    const ctx = await setup();
    ctx.geometry.failWith({
      code: "NotClosed",
      message: "room not closed",
      entityIds: [ctx.storeyId],
    });
    const result = await ctx.addWall.execute({
      documentId: "doc_1",
      storeyId: ctx.storeyId,
      wallId: "wall_s",
      start: { x: 0, y: 0 },
      end: { x: 4000, y: 0 },
      thicknessMm: 200,
      heightMm: 2800,
      kind: "exterior",
    });
    expect(result.rebuild).toEqual({
      ok: false,
      fault: {
        code: "NotClosed",
        message: "room not closed",
        entityIds: [ctx.storeyId],
      },
    });
  });
});

describe("document reconstitution", () => {
  it("loads SceneIR back into FloorPlanDocument", () => {
    const doc = FloorPlanDocument.create({ id: "doc_x" });
    doc.addWall({
      storeyId: doc.storeys[0]!.id,
      start: PointMm.of(0, 0),
      end: PointMm.of(1000, 0),
      thickness: LengthMm.of(120),
      height: LengthMm.of(2800),
      kind: "interior",
    });
    const restored = FloorPlanDocument.fromSceneIR(doc.toSceneIR());
    expect(restored.storeys[0]?.walls).toHaveLength(1);
  });
});

File: core/tests/handlers_test.cpp
#include <gtest/gtest.h>

#include "toporoom/adapters/fake_geometry_port.hpp"
#include "toporoom/adapters/in_memory_document_store.hpp"
#include "toporoom/app/add_opening_handler.hpp"
#include "toporoom/app/add_wall_handler.hpp"
#include "toporoom/app/document_io.hpp"
#include "toporoom/app/set_measurement_handler.hpp"
#include "toporoom/domain/events.hpp"
#include "toporoom/domain/floor_plan_document.hpp"
#include "toporoom/ports/geometry_port.hpp"

using toporoom::adapters::FakeGeometryPort;
using toporoom::adapters::InMemoryDocumentStore;
using toporoom::app::AddOpeningCommand;
using toporoom::app::AddOpeningHandler;
using toporoom::app::AddWallCommand;
using toporoom::app::AddWallHandler;
using toporoom::app::GeometryRebuildPolicy;
using toporoom::app::SetMeasurementCommand;
using toporoom::app::SetMeasurementHandler;
using toporoom::domain::CreateFloorPlanProps;
using toporoom::domain::event_type;
using toporoom::domain::FloorPlanDocument;
using toporoom::domain::MeasurementSource;
using toporoom::domain::OpeningKind;
using toporoom::domain::WallKind;
using toporoom::ports::FaultCode;

namespace {

struct Ctx {
  InMemoryDocumentStore store;
  FakeGeometryPort geometry;
  GeometryRebuildPolicy rebuild;
  AddWallHandler add_wall;
  AddOpeningHandler add_opening;
  SetMeasurementHandler set_measurement;
  std::string storey_id;

  Ctx()
      : rebuild(geometry),
        add_wall(store, rebuild),
        add_opening(store, rebuild),
        set_measurement(store, rebuild) {
    auto doc = FloorPlanDocument::create(CreateFloorPlanProps{"doc_1"});
    storey_id = doc.storeys()[0].id();
    store.save(doc.to_scene_ir());
  }
};

AddWallCommand south_wall(const std::string& storey_id) {
  AddWallCommand cmd;
  cmd.document_id = "doc_1";
  cmd.storey_id = storey_id;
  cmd.wall_id = "wall_s";
  cmd.end_x = 4000;
  cmd.thickness_mm = 200;
  cmd.height_mm = 2800;
  cmd.kind = WallKind::Exterior;
  return cmd;
}

}  // namespace

TEST(Handlers, AddWallMutatesAndRebuilds) {
  Ctx ctx;
  const auto result = ctx.add_wall.execute(south_wall(ctx.storey_id));
  bool saw_wall = false;
  for (const auto& event : result.events) {
    if (std::string(event_type(event)) == "WallAdded") saw_wall = true;
  }
  EXPECT_TRUE(saw_wall);
  EXPECT_TRUE(result.rebuild.ok);
  ASSERT_NE(ctx.geometry.last_build_request(), nullptr);
  EXPECT_EQ(ctx.geometry.last_build_request()->semantics.document_id, "doc_1");
  const auto saved = ctx.store.load("doc_1");
  ASSERT_TRUE(saved);
  EXPECT_EQ(saved->storeys[0].walls.size(), 1u);
}

TEST(Handlers, AddOpeningEmitsEvent) {
  Ctx ctx;
  ctx.add_wall.execute(south_wall(ctx.storey_id));
  AddOpeningCommand opening;
  opening.document_id = "doc_1";
  opening.storey_id = ctx.storey_id;
  opening.wall_id = "wall_s";
  opening.opening_id = "op_door";
  opening.kind = OpeningKind::Door;
  opening.width_mm = 900;
  opening.height_mm = 2100;
  opening.offset_mm = 800;
  const auto result = ctx.add_opening.execute(opening);
  bool saw = false;
  for (const auto& event : result.events) {
    if (std::string(event_type(event)) == "OpeningAdded") saw = true;
  }
  EXPECT_TRUE(saw);
}

TEST(Handlers, SetMeasurementWritesLaserAndResizes) {
  Ctx ctx;
  ctx.add_wall.execute(south_wall(ctx.storey_id));
  AddOpeningCommand opening;
  opening.document_id = "doc_1";
  opening.storey_id = ctx.storey_id;
  opening.wall_id = "wall_s";
  opening.opening_id = "op_door";
  opening.kind = OpeningKind::Door;
  opening.width_mm = 800;
  opening.height_mm = 2100;
  opening.offset_mm = 800;
  ctx.add_opening.execute(opening);
  SetMeasurementCommand measurement;
  measurement.document_id = "doc_1";
  measurement.measurement_id = "m_door";
  measurement.value_mm = 900;
  measurement.source = MeasurementSource::Laser;
  measurement.instrument_id = "laser_sku_x";
  measurement.target = {{"opening", "op_door", "width"}};
  ctx.set_measurement.execute(measurement);
  const auto saved = ctx.store.load("doc_1");
  ASSERT_TRUE(saved);
  EXPECT_EQ(saved->measurements[0].source, MeasurementSource::Laser);
  EXPECT_EQ(saved->measurements[0].value_mm, 900);
  EXPECT_EQ(saved->storeys[0].walls[0].openings[0].width_mm, 900);
}

TEST(Handlers, RebuildPolicyReceivesFault) {
  Ctx ctx;
  ctx.geometry.fail_with({FaultCode::NotClosed, "room not closed", {ctx.storey_id}});
  const auto result = ctx.add_wall.execute(south_wall(ctx.storey_id));
  EXPECT_FALSE(result.rebuild.ok);
  EXPECT_EQ(result.rebuild.fault.code, FaultCode::NotClosed);
}
